### ISSUE TITLE
Missing docs lint ineffectual in test builds

### ISSUE THREAD

**Issue by fenhl** [rust-lang/rust#24584]
The `missing_docs` lint does not do anything in test builds. Minimal code example:

```rust
#![cfg_attr(test, deny(missing_docs))]
```

To reproduce, make a new crate, replace `src/lib.rs` with the above code, and run `cargo test`. The command will run successfully. I would expect the command to report a “missing documentation for crate” compile error.

As for the cause, my wild guess is some sort of interaction between `cfg(test)` and which items are detected as public.

**Comment by steveklabnik** [comment]
Triage: as a slightly more full example, this crate will compile with `--test`, but not compile without it:

```
#![deny(missing_docs)]

#[test]
fn foo() {}

fn lol() {}

fn main() {}
```

**Comment by steveklabnik** [comment]
Triage: no change

**Comment by ehuss** [comment]
This is also an issue for editors that use `cargo check --profile=test` in order to include tests in the check. This means the lint doesn't ever fire in the editor.

It looks like the check was added in https://github.com/rust-lang/rust/commit/4a5d887b58ff9833a968e7a0d28282b915e01de8.  I suspect the reasoning is that code inside a `cfg(test)` block should not be checked for missing docs, which makes sense.  I think it would be nicer to allow missing_docs to run with `--test`, but to not check any block/module/item that is marked with `cfg(test)`. Unfortunately, I think that is very difficult or impossible to do. 😦
